src/memkind_memtier.c

```diff
@@ -1105,14 +1105,13 @@ MEMKIND_EXPORT void *memtier_realloc(struct memtier_memory *memory, void *ptr,
 MEMKIND_EXPORT void *memtier_kind_realloc(memkind_t kind, void *ptr,
                                           size_t size)
 {
-    size_t old_size = 0u;
-    if (size == 0 && ptr != NULL) {
-        old_size = jemk_malloc_usable_size(ptr);
 #ifdef MEMKIND_DECORATION_ENABLED
         if (memtier_kind_free_pre)
             memtier_kind_free_pre(&ptr);
 #endif
 
+    size_t old_size = jemk_malloc_usable_size(ptr);
+    if (size == 0 && ptr != NULL) {       
```

Comment:
This is the *if* block where macro from line 1111 (old version)/ 1108 (new version) should be called.

We use realloc with size *0* instead of free and, in this case, this is the conditional block that we enter. Free decorators should go here.